### DISCUSSION TITLE
Which files / classes to modify if we want to modify negotiation module ?

### DISCUSSION THREAD

**Discussion by SyedMuhamadYasir** [yasserfarouk/scml#39] [standard-qa] [answered]
We want to do implement or modify **three things** in the negotiation approach of our agent ;

- **Bidding Strategy**
- **Opponent Modeling**
- **Acceptance Strategy**

I believe this will entail changing the code of the classes or methods of SCML2020 source files ?

can you please give a little guidance on what exactly should we change ? which **files / classes / methods** would be involved ?

We would be earnestly grateful for a helpful and prompt reply 

Thanks very much !

**Comment by yasserfarouk** [comment, marked as answer]
> can you please give a little guidance on what exactly should we change ? which files / classes / methods would be involved ?

Thank you for your interest in SCML.

The answer depends on the track you will want to participate in.

For the Oneshot track, the class you need to *inherit from* and implement is **OneshotAgent**. You can find a tutorial with several examples [here](http://www.yasserm.com/scml/scml2020docs/tutorials/02.develop_agent_scml2020_oneshot.html#building-your-own-agent). You can also find a skeleton to start your development with methods that need to be implemented  marked [here](http://www.yasserm.com/scml/oneshot.zip). If you are using the skeleton, then you need to modify the methods of `MyAgent` class in `myagent.py`.

You need to implement at least two methods: `propose()` and `respond()`. 

In terms of your question:

- Bidding strategy = `propose()`
- Acceptance strategy = `respond()`
- Opponent model. You can use `step()` and `before_step()` for that.


For the standard and collusion tracks, you need more components because your agent needs to also do long-term planning. 
We have two skeletons that you can use:

1. A monolithic agent skeleton which implements everything (except negotiation details) in a single class. A skeleton is available [here](http://www.yasserm.com/scml/std_monolithic.zip) with all methods that can be implemented marked. If you are going to use this route, then you need to implement another class that does the negotiation for you or use one of the built-in types in negmas/genius. It should be of type `SAONegotiator` or `SAOController`.   You can find an example of using this approach in out tutorials [here](http://www.yasserm.com/scml/scml2020docs/tutorials/03.develop_agent_scml2020.html#an-agent-from-scratch). Again in terms of your question:
    - Bidding strategy = `propose()`
    - Acceptance strategy = `respond()`
    - Opponent model. You can use `step()` and `before_step()` for that.

2. A component based approach in which you reuse some of the built-in components and develop only the ones you need to modify. A skeleton is available [here](http://www.yasserm.com/scml/std_components.zip). In this case, the strategies you are speaking about are part of the `NegotiationManager`. So you need to implement a new one for your specific strategies. You can find an example of this approach in our tutorials [here](http://www.yasserm.com/scml/scml2020docs/tutorials/03.develop_agent_scml2020.html#changing-the-negotiation-strategy). You can find other examples of negotiation managers [here](https://github.com/yasserfarouk/scml/blob/master/src/scml/scml2020/components/negotiation.py). Again, you will need to implement either a `SAONegotiator` or `SAOController` class for your strategies or use one of negmas/genius built in negotiators. 

If you have any other questions, please do not hesitate to contact us. 
You can find step by step directions about the competition on the official website [https://scml.cs.brown.edu](https://scml.cs.brown.edu) and the video/text tutorials linked therein.
